<?xml version="1.0"?>
<robot xmlns:xacro="http://ros.org/wiki/xacro" name="car_trailer_robot">

  <xacro:macro name="wheel" params="name parent_link origin_xyz origin_rpy axis_xyz">
    <link name="${name}">
      <visual>
        <geometry>
          <cylinder radius="0.27" length="0.05"/>
        </geometry>
        <material name="black">
          <color rgba="0 0 0 1"/>
        </material>
      </visual>
      <collision>
        <geometry>
          <cylinder radius="0.27" length="0.05"/>
        </geometry>
      </collision>
      <inertial>
        <mass value="1"/>
        <inertia ixx="0.01" ixy="0" ixz="0" iyy="0.01" iyz="0" izz="0.01"/>
      </inertial>
    </link>

    <joint name="${name}_joint" type="continuous">
      <parent link="${parent_link}"/>
      <child link="${name}"/>
      <origin xyz="${origin_xyz}" rpy="${origin_rpy}"/>
      <axis xyz="${axis_xyz}"/>
    </joint>
  </xacro:macro>

  <!-- Car Base Link -->
  <link name="car_base">
    <visual>
      <geometry>
        <box size="1.3 2 1"/>
      </geometry>
      <material name="blue">
        <color rgba="0 0 1 1"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <box size="1.3 2 0.1"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="10"/>
      <inertia ixx="0.1" ixy="0" ixz="0" iyy="0.1" iyz="0" izz="0.1"/>
    </inertial>
  </link>

  <!-- Car Wheels -->
  <xacro:wheel name="car_wheel_fl" parent_link="car_base" origin_xyz="-0.65 0.6 -0.5" origin_rpy="0 -1.57070 0" axis_xyz="0 0 -1"/>
  <xacro:wheel name="car_wheel_fr" parent_link="car_base" origin_xyz="0.65 0.6 -0.5" origin_rpy="0 1.5707 0" axis_xyz="0 0 1"/>
  <xacro:wheel name="car_wheel_rl" parent_link="car_base" origin_xyz="-0.65 -0.6 -0.5" origin_rpy="0 -1.57070 0" axis_xyz="0 0 -1"/>
  <xacro:wheel name="car_wheel_rr" parent_link="car_base" origin_xyz="0.65 -0.6 -0.5" origin_rpy="0 1.5707 0" axis_xyz="0 0 1"/>

  <!-- Trailer Base Link -->
  <link name="trailer_base">
    <visual>
      <geometry>
        <box size="1.44 2.4 1"/>
      </geometry>
      <material name="red">
        <color rgba="1 0 0 1"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <box size="1.44 2.4 1"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="15"/>
      <inertia ixx="0.15" ixy="0" ixz="0" iyy="0.15" iyz="0" izz="0.15"/>
    </inertial>
  </link>

  <!-- Trailer Connector Link -->
  <link name="trailer_connector_base">
    <visual>
      <geometry>
        <box size="0.1 1.3 0.1"/>
      </geometry>
      <material name="cyan">
        <color rgba="1 1 1 1"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <box size="0.1 1.3 0.1"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="0.1"/>
      <inertia ixx="0.1" ixy="0" ixz="0" iyy="0.1" iyz="0" izz="0.1"/>
    </inertial>
  </link>

  <!-- Trailer Wheels -->
  <xacro:wheel name="trailer_wheel_fl" parent_link="trailer_base" origin_xyz="-0.72 0.72 -0.5" origin_rpy="0 -1.57070 0" axis_xyz="0 0 -1"/>
  <xacro:wheel name="trailer_wheel_fr" parent_link="trailer_base" origin_xyz="0.72 0.72 -0.5" origin_rpy="0 1.5707 0" axis_xyz="0 0 1"/>
  <xacro:wheel name="trailer_wheel_rl" parent_link="trailer_base" origin_xyz="-0.72 -0.72 -0.5" origin_rpy="0 -1.57070 0" axis_xyz="0 0 -1"/>
  <xacro:wheel name="trailer_wheel_rr" parent_link="trailer_base" origin_xyz="0.72 -0.72 -0.5" origin_rpy="0 1.5707 0" axis_xyz="0 0 1"/>

  <!-- Joint between Car base and Trailer connector -->
  <joint name="car_trailer_connector_joint" type="revolute">
    <parent link="car_base"/>
    <child link="trailer_connector_base"/>
    <origin xyz="0 -1 0" rpy="0 0 0"/>
    <axis xyz="0 0 1"/>
    <limit effort="1000" lower="-1.1" upper="1.1" velocity="1"/>
  </joint>

  <!-- Joint between Trailer connector and Trailer -->
  <joint name="car_trailer_joint" type="fixed">
    <parent link="trailer_connector_base"/>
    <child link="trailer_base"/>
    <origin xyz="0 -1.8 0" rpy="0 0 0"/>
    <axis xyz="0 0 1"/>
    <limit effort="1000" lower="-3.14159" upper="3.14159" velocity="1"/>
  </joint>

</robot>

